### PR TITLE
Use trailing commas to allow for simpler refactors

### PR DESCRIPTION
**Pull request by adkron**
Using a trailing comma in a hash or array allows for items in the structure
to be arranged without worrying about the comma. The last item can be
removed without the worry of the comma above still being there.

When looking at a PR there will be one less line changed when adding/removing
an item. This will allow the PR to have more focus and less mental cognitive noise.

Amos King @adkron <amos@binarynoggin.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2718,24 +2718,33 @@ Translations of the guide are available in the following languages:
...
+  # good - easier to move/add/remove items
   VALUES = [
              1001,
              2020,
              3333,
            ]
 
-  # bad
+  # ok
   VALUES = [1001, 2020, 3333, ]
```
Comment (rnkaufman) [inline README.md]: I like the trailing comma in cases like the one above where every value is on a new line (since it helps with "when looking at a PR there will be one less line changed when adding/removing
an item," but I don't think the trailing comma makes sense in this case. In this case, you will have the entire line showing up as changed anyway and I think this looks a little odd.

Comment (adkron) [inline README.md]: That is why I said `ok` instead of good, but I probably shouldn't have made the one below say `bad`. It still makes change simpler by having a trailing comma, but it makes less difference. If you want to expand to multiline it is one less change.

**Comment by adkron** [comment]
I was actually intrigued that the original talked about the advantages of a trailing comma, but then suggested not to use them. It never discussed any advantages to not using them. It also didn't discuss disadvantages of using them.

File: README.md (hunk 1 above)
Comment (rnkaufman) [inline README.md]: Yeah, the labeling of `bad` for the example without the trailing comma was what prompted me to raise this point

**Comment by adkron** [comment]
Thanks for fixing my typos, @messanjah! :+1+

**Comment by HParker** [comment]
Do we feel ok with both

```[1001, 2020, 3333]```
and
```
[
  1,
  2,
  3,
]
```
 being good?

I don't have a problem with it, but I could see it being confusing and seem more inconsistent.

**Comment by siggymcfried** [comment]
Because you asked the nay votes why:

To me, most of my stylistic preferences come down to code being read more often than written. It seems like most agree that `[1, 2, 3,]` seems odd - like an incomplete array. Where we disagree is around whether
```ruby
[
  1,
  2,
  3,
]
```
has the same problem. To me it still reads like an array that is missing an element. I can get past a one line comma addition pretty easily during CR, but it takes me more time to get over the surprise of an array that appears incomplete.

That's a surprise that comes every time I read the code, rather than a surprise once in CR.

**Comment by AllPurposeName** [comment]
Here are my two cents:

Pros:
* Diff looks slightly prettier.
* Ease of development. I recently worked with a spec involving 5 different large arrays. When moving elements to different arrays, I could just plug them in and not need to worry about their former position. I ended up alphabetizing anyway, so the effort saved was minute, really.
* Treating Arrays like linked lists. The tail of a non-circular linked list references nil. Assuming Nothing is Something™, the trailing comma IS still a separator for the "last" element and the nil which inevitably comes after it.
```ruby
stuff = [0,1,2,]
stuff[3] #=> nil
```
Cons:
* Code looks slightly less pretty.
* Not hard at all to fix up your Arrays (even during active development) so the last element has no comma.
* Principle of least surprise can be violated when similar cases of lists behave differently. Though an argument list and an Array are fundamentally different, they look somewhat the same. It requires extra work on my part as a reader to stop and identify what list I am looking at.
```ruby
attr_method :abc,
            :def,
            :hij # Errors with comma on this line

attr_array = [
              :abc,
              :def,
              :hij # Fine with or without comma
             ]
```


tldr; no practical motivation to change, maybe practical motivation not to change. 👎

**Comment by adkron** [comment]
@AllPurposeName 

~~~ruby
irb(main):001:0> stuff = [0,1,2]
=> [0, 1, 2]
irb(main):002:0> stuff.size
=> 3
irb(main):003:0> stuff[3]
=> nil

irb(main):004:0> trailing = [1,2,3,]
=> [1, 2, 3]
irb(main):005:0> trailing[3]
=> nil
irb(main):006:0> trailing.size
=> 3
irb(main):007:0> 
~~~

The two behave the same.

**Comment by AllPurposeName** [comment]
Yes they behave the same, but the latter does not hint that there is a nil element after the 3.

**Comment by HParker** [comment]
```irb(main):003:0> foo = 1,2,3
=> [1, 2, 3]
irb(main):004:0> foo = 1,2,3,
irb(main):005:0* puts "done"
SyntaxError: (irb):5: syntax error, unexpected tSTRING_BEG, expecting keyword_do or '{' or '('
puts "done"
      ^
	from /usr/local/rbenv/versions/2.3.1/bin/irb:11:in `<main>'```

I think this is what @AllPurposeName is talking about.

**Comment by adkron** [comment]
@HParker, I agree that the second part in the cons did say that, but his first example was about the array itself.

**Comment by adkron** [comment]
@AllPurposeName, neither hint that there is a `array[1000]` that is `nil` either. Although it really isn't `nil` all the time. 

~~~ruby
a = Array.new

**Comment by adkron** [comment]
We let this simmer and the voting is 7 to 3. Looks like we have a winner.

**Comment by HParker** [comment]
@adkron, when was this vote?

**Comment by adkron** [comment]
We asked in the channel and there were thumbs up and down at the top. The ticket got quiet and was 7 to 3. Thought that was a good sign that it was done. So I merged.

Amos King
Binary Noggin
